controller/src/main/java/io/pravega/controller/server/retention/StreamCutBucketService.java

```diff
@@ -125,18 +125,19 @@ protected void doStart() {
     }
 
     private CompletableFuture<Void> performRetention(StreamImpl stream) {
-        log.debug("Periodic background processing for retention called for stream {}/{}", stream.getScope(), stream.getStreamName());
         OperationContext context = streamMetadataStore.createContext(stream.getScope(), stream.getStreamName());
+        long requestId = System.nanoTime();
+        log.debug("[requestId={}] Periodic background processing for retention called for stream {}/{}", requestId, stream.getScope(), stream.getStreamName());
```

Comment:
This `[requestId={}]` shows up in every single line of logging code. Is there any way we can reduce the copy-pasting (and remembering) required for this?
- Is there a feature in Slf4J that can use?
- Could we extract this into a shared method in `LoggerHelpers` that will do this for us - but which will not add any additional overhead (such as string concatenation, etc)?

Reply:
I refactored all the locations where `[requestId={}]` was added. Instead, now there are helper methods in `LoggerHelpers` (e.g., `infoLogWithTag`) that do that for you. However, for enabling a user to provide both a `message` and some `args` to be set by Slf4j to that message I couldn't avoid to concatenate strings:  `log.info("[requestId={}] ".concat(message), requestId, args)`.

In this sense, there is the `MDC` mechanism in Slf4j that helps to do a similar things, but the problem is that `MDC` uses thread-local variables to store items to be logged later on. While this works when a single thread is in charge of a well-defined piece of work (e.g., connection, transaction), this seems to do not be reliable for thread pools and async tasks (you can see that in the appendix of the PDP).

Therefore, with this change we are improving code clarity at the cost of some performance penalty (extra string concatenation). At the moment, I found two reasons that may justify this change:
1. Request ids are only logged for `stream operations` (e.g., `createStream`, `deleteStream`), so this is not used for appends/reads which are expected to be orders of magnitude more frequent and performance-sensitive.  
2. I executed a small benchmark locally consisting of logging in a loop `100K` lines with and without that string concatenation. The result is that in most cases the performance penalty is around `15-20%`. Also note that this workload is an extreme case to understand the worst case performance penalty, which is not going to happen in reality.